### ISSUE TITLE
Qgis transforming coordinates incorrectly

### ISSUE THREAD

**Issue by kaymes** [qgis/QGIS#51792]
### What is the bug or the crash?

When converting coordinates between EPSG:3857 and EPSG:7856, QGis calculates incorrect values.

### Steps to reproduce the issue

1. Create print layout
2. Create a label
3. Set the following expression as the label caption: `geom_to_wkt(transform(transform(make_point(246868,6315147),'EPSG:3857','EPSG:7856'),'EPSG:7856','EPSG:3857'))`
4. QGis calculates as result as `Point (246867.54584937 6315145.41463204)`

The expression transforms the point from EPSG:3857 to EPSG:7856 and back. So the result should be the original point again. The result that QGis presents is about 1.1 km off the mark, so much more than a negligible rounding error.

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.28.3-Firenze | QGIS code revision | c12bcb2f76
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.10.6
GDAL/OGR version | 3.4.1
PROJ version | 8.2.1
EPSG Registry database version | v10.041 (2021-12-03)
GEOS version | 3.10.2-CAPI-1.16.0
SQLite version | 3.37.2
PDAL version | 2.3.0
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.4
QScintilla2 version | 2.11.6
OS version | Ubuntu 22.04.1 LTS
  |   |   |  
Active Python plugins
compass_routes | 0.3.1
magnetic_bearing_measurement | 0.5
MetaSearch | 0.3.6
db_manager | 0.1.20
grassprovider | 2.12.99
processing | 2.12.99
sagaprovider | 2.12.99

</body></html>QGIS version
3.28.3-Firenze
QGIS code revision
[c12bcb2f76](https://github.com/qgis/QGIS/commit/c12bcb2f76)
Qt version
5.15.3
Python version
3.10.6
GDAL/OGR version
3.4.1
PROJ version
8.2.1
EPSG Registry database version
v10.041 (2021-12-03)
GEOS version
3.10.2-CAPI-1.16.0
SQLite version
3.37.2
PDAL version
2.3.0
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.4
QScintilla2 version
2.11.6
OS version
Ubuntu 22.04.1 LTS

Active Python plugins
compass_routes
0.3.1
magnetic_bearing_measurement
0.5
MetaSearch
0.3.6
db_manager
0.1.20
grassprovider
2.12.99
processing
2.12.99
sagaprovider
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

**Comment by agiudiceandrea** [comment]
@kaymes, I cannot reproduce the issue using QGIS 3.28.3 on Windows. Could you please try if the issue does occur also using a new QGIS user profile without non core plugin installed?
Anyway, the original point coordinates are 246868, 6315147, and the transformed and transformed back point coordinates are 246867.54584937, 6315145.41463204, so not 1.1 km apart from the original point, but about 2 metre (although even this should not occur).

**Comment by kaymes** [comment]
@agiudiceandrea, I just tried with a new profile and deactivated all plugins. The bug persists and I'm still getting the same result of `Point (246867.54584937 6315145.41463204)`. 

Yes, you are correct, I mixed up scales, so it's m, not km. This makes me wonder: EPSG:7856 (the UTM zone 56S using a datum that's anchored on Australia) uses a geoid that differs by something in the order of two meters from the common WGS 84 datum. Maybe incorrect handling of the datum could be the issue.

**Comment by kaymes** [comment]
I just tried to do the transformation with `gdaltransform` directly and got the same result. So if you've got a more recent version of Gdal, the problem is probably something in the version of Gdal that currently ships with Ubuntu.

```
$ gdaltransform -s_srs EPSG:3857 -t_srs EPSG:7856
Enter X Y [Z [T]] values separated by space, and press Return.
246868 6315147
-1609451.21300748 24117706.8190226 0
$ gdaltransform -s_srs EPSG:7856 -t_srs EPSG:3857
Enter X Y [Z [T]] values separated by space, and press Return.
-1609451.21300748 24117706.8190226
246867.545849343 6315145.41463199 0
$ gdaltransform --version
GDAL 3.4.1, released 2021/12/27
kaymes@durian:~$ 
```

**Comment by agiudiceandrea** [comment]
@kaymes I think the issue is more likely a PROJ issue rather then a GDAL or QGIS issue. Could you please check with the PROJ `cs2cs` tool with `cs2cs "EPSG:3857" "EPSG:7856"` and then enter the coordinates and using the resulting coordinates with `cs2cs "EPSG:7856" "EPSG:3857EPSG:7856"`?

**Comment by kaymes** [comment]
@agiudiceandrea, I just did the test with cs2cs. I'm getting wrong results as well.

```
$ cs2cs "EPSG:3857" "EPSG:7856"
246868 6315147
-1609451.21	24117706.82 0.00
$ cs2cs "EPSG:7856" "EPSG:3857"
-1609451.21	24117706.82
246867.54	6315145.42 0.00
$ cs2cs
Rel. 8.2.1, January 1st, 2022

**Comment by agiudiceandrea** [comment]
@kaymes, it seems to me this would confirm that the issue is in the PROJ library and transformation database distributed in your system. Probably @rouault may confirm or reject this assumption.

**Comment by rouault** [comment]
> The point is in Australia near Sydney, and lies within the area where EPSG:7856 is valid.

not exactly. 246868 6315147 in EPSG:3857 is in France, not very far from where I leave actually. Australia is in the Southern Hemisphere so northings should be negative at the very least...

But, besides that little detail, I do reproduce this missed roundtripping with PROJ 8.2.1, but not with 9.2dev. That said, that area is a bit fragile, and it could potentially fail in the future, because there's unfortunately no guarantee that looking for transformations from B -> A gives the same as the inverse of the ones of A -> B

**Comment by kaymes** [comment]
Thanks for looking into this. It's probably a fixed proj bug. Unfortunately there seems to be no easy way of updating the package without having to also deal with a whole lot of packages that depend on it - but I guess that's my problem, not yours.

@rouault, sorry about the coordinates - it looks like I accidentally used the EPSG:7856 numbers for EPSG:3857 when I constructed the failing example for the bug report.

**Comment by agiudiceandrea** [comment]
@kaymes is the `geom_to_wkt(transform(transform(make_point(246868,6315147),'EPSG:7856','EPSG:3857'),'EPSG:3857','EPSG:7856'))` output also incorrect?

**Comment by kaymes** [comment]
> @kaymes is the `geom_to_wkt(transform(transform(make_point(246868,6315147),'EPSG:7856','EPSG:3857'),'EPSG:3857','EPSG:7856'))` output also incorrect?

Yes, I'm getting `'Point (246868.48850085 6315148.43021531)'`
